build.rs

```diff
@@ -757,8 +762,36 @@ fn configure(interpreter_config: &InterpreterConfig) -> Result<()> {
         _ => {}
     }
 
-    if interpreter_config.shared {
-        println!("cargo:rustc-cfg=Py_SHARED");
+    if env::var_os("CARGO_FEATURE_AUTO_INITIALIZE").is_some() {
+        ensure!(
+            interpreter_config.shared,
+            "The `auto-initialize` feature is enabled, however your python installation only \
+            supports embedding the Python interpreter statically. This is most often seen in \
+            Python distributions compiled from source without the `--enable-shared` configuration \
+            option. For example, this is the default for `pyenv` on macOS.\n\
+            \n\
+            Embedding the Python interpreter statically has many complications and does not yet \
+            have first-class support in PyO3. If you are sure you intend to do this, disable the \
+            `auto-initialize` feature and call `pyo3::prepare_freethreaded_python` directly in \
+            your program.\n\
```

Comment:
I feel this a bit redundant. Users just need to know it's a rare usage and not supported very well, like: `Or, if you are sure to embed the Python interpreter statically (it's a rare usage and not supported very well, though), please ...`